### PR TITLE
[SYCL] Add bfloat16 'hello world' host test.

### PR DESCRIPTION
**Pull request by kbobrovs**
Test for https://github.com/intel/llvm/pull/6675.
Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

### CODE REVIEW

File: SYCL/BFloat16/bfloat_hw.cpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
```
Comment (AlexeySachkov) [inline SYCL/BFloat16/bfloat_hw.cpp]:
```suggestion
```

Host device is going to be removed soon. Moreover, you don't even use any devices in the test, so no need for any special placeholders

**Review by AlexeySachkov — DISMISSED** [review]

File: SYCL/BFloat16/bfloat_hw.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    } else {
+      std::cout << "failed. " << Val << "(Gold) != " << NewVal << "\n";
+    }
+  }
+  return passed;
+}
+
+int main() {
+  bool passed = true;
+  passed &= test(3.140625, 0x4049);
```
Comment (aelovikov-intel) [inline SYCL/BFloat16/bfloat_hw.cpp]:
```suggestion
  passed &= test(3.140625f, 0x4049);
```

File: SYCL/BFloat16/bfloat_hw.cpp (hunk 1 above)
Comment (aelovikov-intel) [reply SYCL/BFloat16/bfloat_hw.cpp]: Maybe add a `single_task` to verify it works on device as well then?

Comment (kbobrovs) [reply SYCL/BFloat16/bfloat_hw.cpp]: Should I just say
`// RUN: %t.out`
?

Comment (kbobrovs) [inline SYCL/BFloat16/bfloat_hw.cpp]:
```suggestion
// RUN: %t.out
```

Comment (kbobrovs) [reply SYCL/BFloat16/bfloat_hw.cpp]: > Maybe add a single_task to verify it works on device as well then?

Yes, good idea. But this is E2E regression test for particular problem I found on host. Test coverage can be improved in separate patches, I believe.

**Comment by asudarsa** [comment]
Adding @rdeodhar who is working on moving bfloat16 support out of experimental. Rajiv,
can you please take a look? Thanks

**Comment by rdeodhar** [comment]
We're moving towards removing the to_float and from_float public interfaces. Also, raw is being removed. This will check the conversions:
```
bfloat16 bf{3.140625f};
if (bf.float() == 3.140625f) then "pass"
```

**Comment by kbobrovs** [comment]
> We're moving towards removing the to_float and from_float public interfaces. Also, raw is being removed. This will check the conversions:
> 
> ```
> bfloat16 bf{3.140625f};
> if (bf.float() == 3.140625f) then "pass"
> ```

I use neither to_float nor from_float - ?

**Comment by kbobrovs** [comment]
... but I use raw(). OK will try to get away only with constructors and bitcasts

**Comment by kbobrovs** [comment]
@rdeodhar, @aelovikov-intel, @AlexeySachkov - please review

File: SYCL/BFloat16/bfloat_hw.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        std::conditional_t<Size == 4, uint32_t,
+                           std::conditional_t<Size == 8, uint64_t, void>>>>;
+
+using bfloat16 = sycl::ext::oneapi::experimental::bfloat16;
+using Bfloat16StorageT = get_uint_type_of_size<sizeof(bfloat16)>;
+
+bool test(float Val, Bfloat16StorageT Bits) {
+  std::cout << "Value: " << Val << " Bits: " << std::hex << "0x" << Bits
+            << std::dec << "...\n";
+  bool passed = true;
```
Comment (aelovikov-intel) [inline SYCL/BFloat16/bfloat_hw.cpp]:
```suggestion
  bool Passed = true;
```

File: SYCL/BFloat16/bfloat_hw.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %t.out
```
Comment (aelovikov-intel) [inline SYCL/BFloat16/bfloat_hw.cpp:2]: I'm not sure if that is what we're supposed to do, but I honestly don't know.

Maybe someone else from @intel/llvm-reviewers-runtime know.

Comment (againull) [reply SYCL/BFloat16/bfloat_hw.cpp:2]: Seems fine to me.

File: SYCL/BFloat16/bfloat_hw.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+bool test(float Val, Bfloat16StorageT Bits) {
+  std::cout << "Value: " << Val << " Bits: " << std::hex << "0x" << Bits
+            << std::dec << "...\n";
+  bool Passed = true;
+  {
+    std::cout << "  float -> bfloat16 conversion ...";
+    Bfloat16StorageT RawVal = sycl::bit_cast<Bfloat16StorageT>(bfloat16(Val));
+    bool Res = (RawVal == Bits);
+    passed &= Res;
```
Comment (kbobrovs) [inline SYCL/BFloat16/bfloat_hw.cpp]:
```suggestion
    Passed &= Res;
```

File: SYCL/BFloat16/bfloat_hw.cpp [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      std::cout << "failed. " << std::hex << "0x" << RawVal << " != "
+                << "0x" << Bits << "(gold)\n"
+                << std::dec;
+    }
+  }
+  {
+    std::cout << "  bfloat16 -> float conversion ...";
+    float NewVal = static_cast<float>(sycl::bit_cast<bfloat16>(Bits));
+    bool Res = (NewVal == Val);
+    passed &= Res;
```
Comment (kbobrovs) [inline SYCL/BFloat16/bfloat_hw.cpp]:
```suggestion
    Passed &= Res;
```

File: SYCL/BFloat16/bfloat_hw.cpp [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+    if (Res) {
+      std::cout << "passed\n";
+    } else {
+      std::cout << "failed. " << Val << "(Gold) != " << NewVal << "\n";
+    }
+  }
+  return passed;
```
Comment (kbobrovs) [inline SYCL/BFloat16/bfloat_hw.cpp]:
```suggestion
  return Passed;
```

File: SYCL/BFloat16/bfloat_hw.cpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+                << std::dec;
+    }
+  }
+  {
+    std::cout << "  bfloat16 -> float conversion ...";
+    float NewVal = static_cast<float>(sycl::bit_cast<bfloat16>(Bits));
+    bool Res = (NewVal == Val);
+    passed &= Res;
+
+    if (Res) {
```
Comment (kbobrovs) [inline SYCL/BFloat16/bfloat_hw.cpp:46]:
```suggestion
    if (Passed) {
```

**Comment by rdeodhar** [comment]
Looks OK.

File: SYCL/BFloat16/bfloat_hw.cpp [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+                << std::dec;
+    }
+  }
+  {
+    std::cout << "  bfloat16 -> float conversion ...";
+    float NewVal = static_cast<float>(sycl::bit_cast<bfloat16>(Bits));
+    bool Res = (NewVal == Val);
+    Passed &= Res;
+
+    if (Passed) {
```
Comment (kbobrovs) [inline SYCL/BFloat16/bfloat_hw.cpp]:
```suggestion
    if (Res) {
```

**Comment by kbobrovs** [comment]
Testing with https://github.com/intel/llvm/pull/6675 has only known TGL-specific failures on Windows

[2022-09-01T22:56:01.536Z]   SYCL :: AtomicRef/assignment_atomic64.cpp
[2022-09-01T22:56:01.536Z]   SYCL :: AtomicRef/assignment_atomic64_generic.cpp
[2022-09-01T22:56:01.536Z]   SYCL :: Basic/buffer/buffer.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: DeprecatedFeatures/SpecConsts1.2.1/specialization_constants.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: DeprecatedFeatures/SpecConsts1.2.1/specialization_constants_override.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: DeviceLib/built-ins/nan.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/aot_mixed.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/api/bin_and_cmp_ops_heavy.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/api/functional/ctors/ctor_converting_fp_extra.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/api/functional/ctors/ctor_load_acc_fp_extra.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/api/functional/ctors/ctor_load_usm_fp_extra.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/api/functional/functions/functions_select_2d_core.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/api/functional/functions/functions_select_2d_fp_extra.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/api/functional/operators/operator_decrement_and_increment_accuracy_fp_extra.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/api/functional/operators/operator_decrement_and_increment_fp_extra.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/api/replicate_smoke.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/api/saturation_smoke.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/api/simd_view_select_2d_fp.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/api/simd_view_select_2d_int.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/api/unary_ops_heavy.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/ext_math.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/matrix_transpose_glb.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/regression/dgetrf.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/regression/dgetrf_8x8.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/regression/dgetrf_ref.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: ESIMD/spec_const/spec_const_double.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: GroupAlgorithm/SYCL2020/sort.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: KernelParams/union_kernel_param.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: SpecConstants/2020/handler-api.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: SpecConstants/2020/kernel-bundle-api.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: SubGroup/barrier.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: SubGroup/broadcast_fp64.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: SubGroup/info.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: SubGroup/load_store.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: SubGroup/shuffle_fp64.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: USM/copy.cpp
[2022-09-01T22:56:01.537Z]   SYCL :: USM/fill.cpp

**Review by AlexeySachkov — APPROVED** [review]

**Comment by kbobrovs** [comment]
bfloat_hw.cpp passed everywhere

**Pull request merged**